### DISCUSSION TITLE
Tailor your EXCD experience

### DISCUSSION THREAD

**Discussion by zer0Kerbal** [zer0Kerbal/ExceptionDetector#42] [FAKQ] [answered]
### Tailor your EXCD experience: 

Is there is a simple and efficient way to stop things from appearing in the EXCD log, other than removing xxx addon?

**Comment by zer0Kerbal** [comment, marked as answer]
Add a line to [GameData/godarklight/ExceptionDetector/Plugins/Settings.cfg)](https://github.com/zer0Kerbal/ExceptionDetector/blob/f8b198571b3d588e0b334d5f7b672b152fbf2473/GameData/godarklight/ExceptionDetector/Plugins/Settings.cfg)

something like:
`kspcf = KSPCF`

I believe that that line will 'filter' out all logging from that mod; and there are other ways to filter that are demonstrated in the settings.cfg

**Comment by Alpaccalypse** [comment]
So, it seems that this isn't a filter function in the way I was hoping. I've spent hours now, using trial and error, to try and work out how this all works, and I've had to give up for now for sanity's sake. It just doesn't seem to work very logically or consistently from my perspective. I have at least manage to come to a few conclusions, so I'll share what I have learned so far, just in case it is not already obvious from your end or might help with future updates:

Firstly I'm not sure KSPCF is the best test case, as EXCD seems to be behaving a oddly with this example, after taking into account what already doesn't work as I expected elsewhere:

- If I add it to as a single pass line as `KSPCOMMINTYFIXES = [KSPCF]` as far as I can tell, absolutely nothing happens (which seems to be a consistent problem when doing a lot of things as a single pass).
- If I add it as a second pass, rather than reporting the exception it causes on the next line, it reports a warning from three lines down saying ...`001_ToolbarControl/PluginData/Debug.cfg' does not exist`.
EXCD also adds this report as a new line in the log,which I assume is WAI despite the wrong error being caught, but I mention this because this behaviour seems inconsistent.

Possibly related to this is that the actual exception reports _above_ the part of the EXCD log which says:
`LogType:	**ED-Exception`
`ED-EXCEPTION****`

What I really wanted to do anyway is stop the log being spammed by so many nearly identical `Condition:	[Part]: Cannot have ModuleCargoPart and ModuleInventoryPart on same Part` messages. The best solution I've found so far is to just remove `PluginLoader:` as a reference from the settings entirely. Though, I have actually left it as a single pass line, just in case this is important. Though it might be significant that absolutely nothing changes in the log or UI report whether it is there or not. I assume this is because the errors only occur as a second log entry.

Unlike [KSPCF] EXCD is reporting the correct errors whether on default settings or with either of the changes mentioned above. Why it is preferable to move it from double to single pass, is that it reports the exact same errors, but excluding the line above. Making the UI log more readable (as every first line is identical, I really don't need it repeated every time) and stopping the log expanding past the bottom of the screen.

Either of these "solutions" stop EXCD logging these errors as a line saying `EXCD	PartLoader: Compiling Part 'Squad/Parts/Cargo/CargoContainers/cargoContainer/cargoContainer'--> [Part]: Cannot have ModuleCargoPart and ModuleInventoryPart on same Part [cargoContainer'. Removed ModuleCargoPart` (with a different name for each entry). It just reports the second half starting with "[PART:]" instead.

What I was really hoping to achieve was preferably being able to "whitelist" these vanilla errors, so they are not reported in the UI at all (and I don't know of any obvious feature to do that) or, nearly as helpful, get them to be listed like the other similar errors, like the wheel collider one which just has a prefix to one error report of "22 times:"

I thought this could be achieved in a similar way that the other double pass lines are used:
`		DoublePass`
`		{`
`			MODELLOADER = Load(Texture):`
`			TEXTURELOADER = Load(Model):`
`			AUDIOLOADER = Load(Audio):`
`		}`

I've spent hours however trying to work out how the syntax works, and I think it might be working as intended and my problem is that whatever is in brackets has to be the first word of the error line.

Using `EXCD	PartLoader: Compiling Part 'Squad/Parts/Cargo/CargoContainers/cargoContainer/cargoContainer'--> [Part]: Cannot have ModuleCargoPart and ModuleInventoryPart on same Part [cargoContainer'. Removed ModuleCargoPart` as an example. I have tried everything I can think of to replicate the functionality of the Load double pass entries, and been unable to change anything at all. I'm not even certain of what requirements are needed or if the options in brackets are hardcoded, I've tried:

- Defining PartLoader: as a single pass and using [Part]: as a second pass
- Every variation of PartLoader/Part and Part/Cannot/Removed in brackets afterwards.
- Tweaking the working Load lines used in the default settings.

The only tangible result I've had is to confirm that those seem to be working exactly as expected with the errors reported, and can be manipulated to a certain degree. As far as I can tell the (Model/Texture/Audio) seems to be the first word of the error line being reported, and that is why it is impossible to get the log to just report "14 times: Removed ModuleCargoPart" or even just give an abbreviated version of each related error message.

Anyway, I was starting to have diminishing returns on my mental capacity from doing this over and over again and thought it was probably more practical to leave an update here and see if you know of anything I am obviously doing wrong, or confirm that this is something that requires further investigation. Also, this has lead me to have a pretty big wishlist of feature requests to post, but I thought I should hold off on that for now, for the same reason. I'll post my current Settings, EXCD log and game log, in case those are helpful. I imagine, at least on some level, I am just being a dumb end user, but either way, hopefully some of this is useful feedback. Apologies if any of this is a bit incoherent, my mental faculties are a bit drained and I wanted to actually write up on this rather than just forget it all later.

[Settings.zip](https://github.com/zer0Kerbal/ExceptionDetector/files/11391061/Settings.zip)
[ed.zip](https://github.com/zer0Kerbal/ExceptionDetector/files/11391065/ed.zip)
[KSP.zip](https://github.com/zer0Kerbal/ExceptionDetector/files/11391070/KSP.zip)

*Edit for code formatting
